client/db/src/lib.rs

```diff
@@ -1504,47 +1500,74 @@ impl<Block: BlockT> Backend<Block> {
 		&self,
 		transaction: &mut Transaction<DbHash>,
 		finalized: NumberFor<Block>,
+		displaced: &FinalizationDisplaced<Block::Hash, NumberFor<Block>>,
 	) -> ClientResult<()> {
 		if let KeepBlocks::Some(keep_blocks) = self.keep_blocks {
 			// Always keep the last finalized block
 			let keep = std::cmp::max(keep_blocks, 1);
-			if finalized < keep.into() {
-				return Ok(())
+			if finalized >= keep.into() {
+				let number = finalized.saturating_sub(keep.into());
+				self.prune_block(transaction, BlockId::<Block>::number(number))?;
 			}
-			let number = finalized.saturating_sub(keep.into());
-			match read_db(&*self.storage.db, columns::KEY_LOOKUP, columns::BODY, BlockId::<Block>::number(number))? {
-				Some(body) => {
-					debug!(target: "db", "Removing block #{}", number);
-					utils::remove_from_db(
-						transaction,
-						&*self.storage.db,
-						columns::KEY_LOOKUP,
-						columns::BODY,
-						BlockId::<Block>::number(number),
-					)?;
-					match self.transaction_storage {
-						TransactionStorageMode::BlockBody => {},
-						TransactionStorageMode::StorageChain => {
-							match Vec::<Block::Hash>::decode(&mut &body[..]) {
-								Ok(hashes) => {
-									for h in hashes {
-										transaction.remove(columns::TRANSACTION, h.as_ref());
-									}
+		}
+		// Also discard all blocks from displaced branches
+		for h in displaced.leaves() {
+			let mut number = finalized;
+			let mut hash = h.clone();
+			// Follow displaced chains up to finality point
```

Comment:
Don't we start at the finalized point and move downwards?

Reply:
Not really. We start with a displaced leaf and follow its parents until we reach a block that is canonical

Reply:
Yeah I understood it this way.

Shouldn't the comment say `Follow displaced chains up to canonical point`?

Reply:
Since they are displaced due to finality, as soon as we meet a canonical parent it is guaranteed to be a part of a finalized chain. I'll update the comment to make it more obvious.